### PR TITLE
[Protocol] Support secrets in protocol

### PR DESCRIPTION
**Pull request by abuccts**
Support secrets in protocol
------------------------------

Users may need sensitive information including password or API key in their jobs, `secrets` section is used to help user hide those sensitive information in plain text and avoid sharing.

Work items
------------
- [x] Agree on an approach to add secrets in protocol.
- [x] Update the protocol implementation to support read secret from yaml and mask with `*` when storing.
- [x] Hide the secrets in webportal, including submission, config view, and job clone.

### CODE REVIEW

File: docs/pai-job-protocol.yaml [hunk 1]
```diff
@@ -19,16 +19,23 @@ prerequisites: # Optional
...
 # If specified, the whole parameters object can be referenced as `$parameters`.
-# Scope of reference `$parameters`: the refercen is global and shared among all task roles.
+# Scope of reference `$parameters`: the reference is global and shared among all task roles.
 parameters: # Optional, can be omitted.
   <param1>: value1 # Specify name and value of all the referencable parameters that will be used in the whole job template.
   <param2>: value2 # Can be referenced by `<% $parameters.param1 %>`, `<% $parameters.param2 %>`.
 
+# If sensitive information including password or API key is needed in the protocol,
+# it should be specified here in secrets section and referenced as `$secrets`.
+# All secrets will be hidden in the plain text and won't be shared.
```
Comment (fanyangCS) [inline docs/pai-job-protocol.yaml]: >All secrets will be hidden in the plain text and won't be shared [](start = 2, length = 64)

A system that supports PAI protocol should not keep information marked as secretes. For example, the yaml file used for job cloning, the stdout/stderr should remove all information marked as secrets.

File: docs/pai-job-protocol.yaml [hunk 2]
```diff
@@ -126,6 +133,10 @@ parameters:
   model: resnet20
   batchsize: 32
 
+secrets:
+  docker_password: password
+  api_token: cGFzc3dvcmQ=
```
Comment (fanyangCS) [inline docs/pai-job-protocol.yaml]: >api_token [](start = 2, length = 9)

where do you use api_token?

Comment (abuccts) [reply docs/pai-job-protocol.yaml]: it hasn't been used in this file, just shows that the secret can be either password or token.

File: docs/pai-job-protocol.yaml (hunk 1 above)
Comment (abuccts) [reply docs/pai-job-protocol.yaml]: updated

File: docs/pai-job-protocol.yaml [hunk 3]
```diff
@@ -19,16 +19,24 @@ prerequisites: # Optional
...
 # If specified, the whole parameters object can be referenced as `$parameters`.
-# Scope of reference `$parameters`: the refercen is global and shared among all task roles.
+# Scope of reference `$parameters`: the reference is global and shared among all task roles.
 parameters: # Optional, can be omitted.
   <param1>: value1 # Specify name and value of all the referencable parameters that will be used in the whole job template.
   <param2>: value2 # Can be referenced by `<% $parameters.param1 %>`, `<% $parameters.param2 %>`.
 
+# If sensitive information including password or API key is needed in the protocol,
+# it should be specified here in secrets section and referenced as `$secrets`.
+# A system that supports PAI protocol should not keep information marked as secretes.
```
[the remark this answers is not in the document]
Comment (abuccts) [reply docs/pai-job-protocol.yaml]: updated

**Comment by fanyangCS** [comment]
> default:

should be "defaults"?

---
Refers to: docs/pai-job-protocol.yaml:235 in 9e65442. [](commit_id = 9e654422c09dff193594b7f9eb313d46227799f7, deletion_comment = False)

**Comment by abuccts** [comment]
> > default:
> 
> should be "defaults"?

right, fixed

File: docs/pai-job-protocol.yaml [hunk 4]
```diff
@@ -19,16 +19,26 @@ prerequisites: # Optional
...
 # If specified, the whole parameters object can be referenced as `$parameters`.
-# Scope of reference `$parameters`: the refercen is global and shared among all task roles.
+# Scope of reference `$parameters`: the reference is global and shared among all task roles.
 parameters: # Optional, can be omitted.
   <param1>: value1 # Specify name and value of all the referencable parameters that will be used in the whole job template.
   <param2>: value2 # Can be referenced by `<% $parameters.param1 %>`, `<% $parameters.param2 %>`.
 
+# If sensitive information including password or API key is needed in the protocol,
+# it should be specified here in secrets section and referenced as `$secrets`.
+# Scope of reference `$secrets`: the reference is global, shared among all task roles and docker image's `auth` field.
```
Comment (fanyangCS) [inline docs/pai-job-protocol.yaml]: "global" means all sections. I think the reference scope is taskRole and auth field.

Comment (abuccts) [reply docs/pai-job-protocol.yaml]: updated

**Pull request merged**